File: e2e-tests/spring-examples/src/test/java/com/foo/rest/examples/spring/SpringController.java
package com.foo.rest.examples.spring;

import org.evomaster.clientJava.controller.EmbeddedSutController;
import org.evomaster.clientJava.controllerApi.dto.AuthenticationDto;
import org.springframework.boot.SpringApplication;
import org.springframework.context.ConfigurableApplicationContext;

import java.sql.Connection;
import java.util.List;
import java.util.Map;

public abstract class SpringController extends EmbeddedSutController {

    protected ConfigurableApplicationContext ctx;
    protected final Class<?> applicationClass;



    protected SpringController(Class<?> applicationClass){
       this.applicationClass = applicationClass;
       super.setControllerPort(0);
    }


    @Override
    public String startSut() {

        ctx = SpringApplication.run(applicationClass, "--server.port=0");


        return "http://localhost:"+getSutPort();
    }

    protected int getSutPort(){
        return (Integer)((Map) ctx.getEnvironment()
                .getPropertySources().get("server.ports").getSource())
                .get("local.server.port");
    }


    @Override
    public boolean isSutRunning() {
        return ctx!=null && ctx.isRunning();
    }

    @Override
    public void stopSut() {
        ctx.stop();
        ctx.close();
    }

    @Override
    public String getPackagePrefixesToCover() {
        return "com.foo.";
    }

    @Override
    public void resetStateOfSUT() {
        //nothing to do
    }

    @Override
    public String getUrlOfSwaggerJSON() {
        return "http://localhost:"+getSutPort()+"/v2/api-docs";
    }

    @Override
    public List<AuthenticationDto> getInfoForAuthentication() {
        return null;
    }

    @Override
    public Connection getConnection() {
        return null;
    }

    @Override
    public String getDatabaseDriverName() {
        return null;
    }

    @Override
    public List<String> getEndpointsToSkip(){
        return null;
    }
}

File: core-tests/e2e-tests/spring/spring-mcp-bb/src/test/kotlin/com/foo/mcp/bb/examples/spring/SpringController.kt
package com.foo.mcp.bb.examples.spring

import org.evomaster.client.java.controller.EmbeddedSutController
import org.evomaster.client.java.controller.api.dto.auth.AuthenticationDto
import org.evomaster.client.java.controller.api.dto.SutInfoDto
import org.evomaster.client.java.controller.problem.GraphQlProblem
import org.evomaster.client.java.sql.DbSpecification
import org.evomaster.client.java.controller.problem.ProblemInfo
import org.springframework.boot.SpringApplication
import org.springframework.context.ConfigurableApplicationContext


abstract class SpringController(protected val applicationClass: Class<*>) : EmbeddedSutController() {


    init {
        super.setControllerPort(0)
    }

    protected var ctx: ConfigurableApplicationContext? = null

    override fun startSut(): String {
        ctx = SpringApplication.run(applicationClass,
            "--server.port=0",
        )
        return "http://localhost:$sutPort"
    }

    val sutPort: Int
        get() = (ctx!!.environment
            .propertySources["server.ports"].source as Map<*, *>)["local.server.port"] as Int

    override fun isSutRunning(): Boolean {
        return ctx != null && ctx!!.isRunning
    }

    override fun stopSut() {
        ctx?.stop()
        ctx?.close()
    }


    override fun getPackagePrefixesToCover(): String {
        return "com.foo.mcp."
    }

    override fun getDbSpecifications(): MutableList<DbSpecification>? {
        return null
    }

    override fun getPreferredOutputFormat(): SutInfoDto.OutputFormat {
        return SutInfoDto.OutputFormat.KOTLIN_JUNIT_5
    }

    override fun resetStateOfSUT() {
    }

    override fun getInfoForAuthentication(): List<AuthenticationDto> {
        return listOf()
    }

    override fun getProblemInfo(): ProblemInfo {
        // TODO change to McpProblem when implemented
        return GraphQlProblem("${getBaseURL()}/graphql")
    }

    fun getBaseURL(): String {
        return "http://localhost:$sutPort"
    }


}
